File: src/main/javascript/crypto/src/libs/FullProofOfExponent.ts
import {CURVE_BN256, Point} from "./Point";
import {Proof} from "../asn1/shemas/ProofOfExponentASN";
import {AsnParser} from "@peculiar/asn1-schema";
import {base64ToUint8array, uint8ToBn, uint8tohex, bnToUint8} from "./utils";
import {Asn1Der} from "./DerUtility";
import {UsageProofOfExponent} from "./UsageProofOfExponent";
import { AttestationCrypto } from "./AttestationCrypto";

export class FullProofOfExponent {
    private riddle: Point;
    private tPoint: Point;
    private challengeResponse: bigint;
    private nonce: Uint8Array;
    public encoding: string;

    private constructor() {}

    static fromData(riddle: Point, tPoint: Point, challengeResponse: bigint, nonce: Uint8Array = new Uint8Array([])) {
        let me = new this();
        me.riddle = riddle;
        me.tPoint = tPoint;
        me.challengeResponse = challengeResponse;
        me.nonce = nonce;
        me.encoding = me.makeEncoding(riddle, tPoint, challengeResponse, nonce);
        return me;
    }

    static fromBytes( uint8data: Uint8Array ) {
        let proof: Proof = AsnParser.parse(uint8data, Proof);
        return this.fromASNType(proof);
    }

    static fromASNType( proof:Proof ) {

        let riddleEnc: Uint8Array = new Uint8Array(proof.riddle);

        let riddle = Point.decodeFromUint8(riddleEnc, CURVE_BN256 );

        let challengeEnc: Uint8Array = new Uint8Array(proof.challengePoint);
        let challengeResponse = uint8ToBn(challengeEnc);

        let tPointEnc: Uint8Array = new Uint8Array(proof.responseValue);
        let tPoint = Point.decodeFromUint8(tPointEnc, CURVE_BN256 );

        let nonce = new Uint8Array(proof.nonce);

        return this.fromData(riddle, tPoint, challengeResponse, nonce);
    }

    static fromBase64(base64DerEncoded: string) {
        return FullProofOfExponent.fromBytes(base64ToUint8array(base64DerEncoded));
    }

    makeEncoding(riddle: Point, tPoint: Point, challengeResponse: bigint, nonce: Uint8Array = new Uint8Array([])):string{
        let proof = Asn1Der.encode('OCTET_STRING', uint8tohex(riddle.getEncoded()))
            + Asn1Der.encode('OCTET_STRING', challengeResponse.toString(16))
            + Asn1Der.encode('OCTET_STRING', uint8tohex(tPoint.getEncoded()))
            + Asn1Der.encode('OCTET_STRING', uint8tohex(nonce));

        return Asn1Der.encode('SEQUENCE_30', proof);
    }

    public getRiddle(): Point {
        return this.riddle;
    }

    public getPoint(): Point {
        return this.tPoint;
    }

    public getChallengeResponse() {
        return this.challengeResponse;
    }

    public getNonce(): Uint8Array {
        return this.nonce;
    }

    public getUsageProofOfExponent(): UsageProofOfExponent {
        return UsageProofOfExponent.fromData(this.tPoint, this.challengeResponse, this.nonce);
    }

    public getDerEncoding(): string {
        return this.encoding;
    }
    
    public getAsnType(): Proof {

        const proof = new Proof();
        proof.nonce = this.getNonce();
        let point = bnToUint8(this.getChallengeResponse());
        if (point.length < 32) {
            let prevPoint = point;
            point = new Uint8Array(32);
            point.set(prevPoint, 32 - prevPoint.length);
        }
        proof.challengePoint = point;
        proof.riddle = this.getRiddle().getEncoded();
        proof.responseValue = this.getPoint().getEncoded();

        return proof;
    }

    public validateParameters():boolean {
        try {
            // Validate that points are valid on the given curve, have correct order and are not at infinity

            // if (!this.riddle.validate() || !this.tPoint.validate()){
            if (
                !AttestationCrypto.validatePointToCurve(this.riddle, AttestationCrypto.curve) 
            || !AttestationCrypto.validatePointToCurve(this.tPoint, AttestationCrypto.curve )
            ){
                throw new Error("Point not in the curve");
            }

            // Check the challenge response size
            if (this.challengeResponse <= 0n || this.challengeResponse >= AttestationCrypto.curveOrder) {
                return false;
            }

            // While not strictly needed also check that points are not the generator
            if (this.riddle.equals( AttestationCrypto.G ) || this.riddle.equals( AttestationCrypto.H ) ) {
                return false;
            }

            if (this.tPoint.equals( AttestationCrypto.G ) || this.tPoint.equals( AttestationCrypto.H )) {
                return false;
            }

            return true;

        } catch ( e ) {
          return false;
        }
    }
}

File: src/main/java/org/tokenscript/attestation/FullProofOfExponent.java
package org.tokenscript.attestation;

import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.ECPublicKeyParameters;
import org.tokenscript.attestation.core.AttestationCrypto;
import org.tokenscript.attestation.core.ExceptionUtil;
import java.io.IOException;
import java.math.BigInteger;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1InputStream;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DEROctetString;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.math.ec.ECPoint;

public class FullProofOfExponent implements ProofOfExponent {
  private static final Logger logger = LogManager.getLogger(FullProofOfExponent.class);
  private final ECPoint riddle;
  private final ECPoint tPoint;
  private final BigInteger challengeResponse;
  private final byte[] unpredictableNumber;
  private final byte[] encoding;

  public FullProofOfExponent(ECPoint riddle, ECPoint tPoint, BigInteger challengeResponse, byte[] unpredictableNumber) {
    this.riddle = riddle;
    this.tPoint = tPoint;
    this.challengeResponse = challengeResponse;
    this.unpredictableNumber = unpredictableNumber;
    this.encoding = makeEncoding(riddle, tPoint, challengeResponse, unpredictableNumber);
  }

  public FullProofOfExponent(ECPoint riddle, ECPoint tPoint, BigInteger challengeResponse) {
    this(riddle, tPoint, challengeResponse, new byte[0]);
  }

  public FullProofOfExponent(byte[] derEncoded) {
    this.encoding = derEncoded;
    try {
      ASN1InputStream input = new ASN1InputStream(derEncoded);
      ASN1Sequence asn1 = ASN1Sequence.getInstance(input.readObject());
      input.close();
      int asn1counter = 0;
      ASN1OctetString riddleEnc = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++));
      this.riddle = AttestationCrypto.decodePoint(riddleEnc.getOctets());
      ASN1OctetString challengeEnc = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++));
      this.challengeResponse = new BigInteger(challengeEnc.getOctets());
      ASN1OctetString tPointEnc = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++));
      this.tPoint = AttestationCrypto.decodePoint(tPointEnc.getOctets());
      this.unpredictableNumber = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++)).getOctets();
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
  }

  private byte[] makeEncoding(ECPoint riddle, ECPoint tPoint, BigInteger challenge, byte[] unpredictableNumber) {
    try {
      ASN1EncodableVector res = new ASN1EncodableVector();
      res.add(new DEROctetString(riddle.getEncoded(false)));
      res.add(new DEROctetString(challenge.toByteArray()));
      res.add(new DEROctetString(tPoint.getEncoded(false)));
      res.add(new DEROctetString(unpredictableNumber));
      return new DERSequence(res).getEncoded();
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not encode asn1", e);
    }
  }

  public ECPoint getRiddle() {
    return riddle;
  }

  @Override
  public ECPoint getPoint() {
    return tPoint;
  }

  @Override
  public BigInteger getChallengeResponse() {
    return challengeResponse;
  }

  @Override
  public byte[] getUnpredictableNumber() { return unpredictableNumber; }

  public UsageProofOfExponent getUsageProofOfExponent() {
    return new UsageProofOfExponent(tPoint, challengeResponse, unpredictableNumber);
  }

  @Override
  public byte[] getDerEncoding() {
    return encoding;
  }

  /**
   * Verify that the proof is correct AND that the values contained are correct according to RFC 8235
   * The latter part is _crucial_ in preventing attacks through edge cases.
   * @return true if everything is ok
   */
  @Override
  public boolean verify() {
    try {
      // Validate that points are valid on the given curve, have correct order and are not at infinity
      AttestationCrypto.validatePointToCurve(riddle, AttestationCrypto.curve);
      AttestationCrypto.validatePointToCurve(tPoint, AttestationCrypto.curve);
      // Check the challenge response size
      if (challengeResponse.compareTo(BigInteger.ZERO) <= 0 || challengeResponse.compareTo(AttestationCrypto.curve.getOrder()) >= 0) {
        return false;
      }
      // While not strictly needed also check that points are not the generator
      if (riddle.equals(AttestationCrypto.G) || riddle.equals(AttestationCrypto.H)) {
        return false;
      }
      if (tPoint.equals(AttestationCrypto.G) || tPoint.equals(AttestationCrypto.H)) {
        return false;
      }
      return AttestationCrypto.verifyFullProof(this);
    } catch (SecurityException e) {
      return false;
    }
  }
}
